frame/evm/precompile/withdraw/src/lib.rs

```diff
@@ -0,0 +1,112 @@
+// This file is part of Darwinia.
+//
+// Copyright (C) 2018-2021 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	ensure,
+	traits::{Currency, ExistenceRequirement},
+};
+use sp_core::U256;
+use sp_runtime::traits::UniqueSaturatedInto;
+use sp_std::marker::PhantomData;
+use sp_std::prelude::*;
+use sp_std::vec::Vec;
+
+use codec::Decode;
+use darwinia_evm::{AddressMapping, Trait};
+use darwinia_evm_primitives::Precompile;
+use evm::{Context, ExitError, ExitSucceed};
+
+type AccountId<T> = <T as frame_system::Trait>::AccountId;
+
+/// WithDraw Precompile Contract, used to withdraw balance from evm account to darwinia account
+///
+/// The contract address: 0000000000000000000000000000000000000005
+pub struct WithDraw<T: Trait> {
+	_maker: PhantomData<T>,
+}
+
+impl<T: Trait> Precompile for WithDraw<T> {
+	/// The Withdraw process is divided into two part:
+	/// 1. parse the withdrawal address from the input parameter and get the contract address and value from the context
+	/// 2. transfer from the contract address to withdrawal address
+	///
+	/// Input data: 32-bit substrate withdrawal public key
+	fn execute(
+		input: &[u8],
+		_: Option<usize>,
+		context: &Context,
+	) -> core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError> {
+		// Decode input data
+		let input = InputData::<T>::decode(&input)?;
+
+		let helper = U256::from(10)
+			.checked_pow(U256::from(9))
+			.unwrap_or(U256::MAX);
+		let contract_address = T::AddressMapping::into_account_id(context.address);
+		let context_value = context.apparent_value.div_mod(helper).0;
+		let context_value = context_value.low_u128().unique_saturated_into();
+
+		let result = T::Currency::transfer(
+			&contract_address,
+			&input.dest,
+			context_value,
+			ExistenceRequirement::AllowDeath,
+		);
+
+		ensure!(
```

Comment:
If support `delegatecall`, there is no guarantee that balance of context.address is zero.

Reply:
Good catch.

Reply:
And if `target` address and `context.address` is same.

Reply:
I see what you mean, it's really not a good idea to check the balance here for a regular contract account.  I will remove this `ensure!`

Reply:
Done